smdebug/core/config_validator.py

```diff
@@ -0,0 +1,106 @@
+from smdebug.core.utils import is_framework_version_supported
+from smdebug.profiler.profiler_config_parser import ProfilerConfigParser
+
+logger = get_logger()
+
+_config_validator = None
+
+
+class FRAMEWORK(Enum):
+    PYTORCH = 0
+    TENSORFLOW = 1
+    MXNET = 2
+    XGBOOST = 3
+
+
+SupportedFrameworks = [FRAMEWORK.PYTORCH, FRAMEWORK.TENSORFLOW, FRAMEWORK.MXNET, FRAMEWORK.XGBOOST]
+
+
+class ConfigValidator(object):
+    def __init__(self, frameworkType: FRAMEWORK):
+        self._create_hook = strtobool(os.getenv("USE_SMDEBUG", "true").lower())
+        self._summary = ""
+        self._frameworkType = frameworkType
+
+    def _validate_training_environment(self):
+        """
+        The function checks whether the debugger functionality can be supported in the given training environment.
+        Currently, if the training job is using a framework that is not supported by the debugger, we disable the
+        creation of hook and disable the debugger functionality.
+        :return:
+        """
+        if (
+            self._frameworkType in SupportedFrameworks
+            and is_framework_version_supported(self._frameworkType) is False
+        ):
+            self._create_hook = False
+
+    @staticmethod
+    def validate_profiler_config(profiler_config_parser: ProfilerConfigParser):
+        """
+        The function parses the profiler configuration and ensures that hte configuration can be supported for the
+        current training job.
+        Currently, if the training job is using smmodel parallel we will disable the autograd profiler
+        :return:
+        """
+        # Since we support the functionality to update profiler config during training, we would need to reload the
+        # configuration and re-evaluate whether we can support autograd profiler.
+        if (
+            profiler_config_parser.profiling_enabled
+            and profiler_config_parser.config.detailed_profiling_config is not None
+        ):
+            if smdebug.core.utils.check_smmodelparallel_training():
+                profiler_config_parser.config.detailed_profiling_config.disabled = True
+                logger.warning("Detailed profiling for model parallel training job is disabled.")
+
+    def _validate_debugger_config(self):
+        """
+        TODO: Analyze the debugger configuration and summarize the performance impact caused by parameters set in
+        collection configuration. A saveall collection or saving tensors for every step wil have a significant
+        performance impact. We can let the user know about this impact through this function.
+        :return:
+        """
+
+    def validate_training_job(self):
+        # If Hook is disabled we need to validate the training job again.
+        if self._create_hook is False:
+            return
+        self._validate_training_environment()
+        self._validate_debugger_config()
+        self.validate_profiler_config(ProfilerConfigParser())
```

Comment:
A profiler config parser object is always created globally in `smdebug/tensorflow/keras.py` (for TF) or `smdebug/pytorch/hook.py` (for PT). Those objects should be passed in as arguments to this function, instead of creating a new profiler config parser object.

https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/tensorflow/keras.py#L55
https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/pytorch/hook.py#L41

Reply:
Example:

```
from smdebug.tensorflow import profiler_config_parser
from smdebug.core.config_validator import FRAMEWORK, get_config_validator
...
get_config_validator(FRAMEWORK.TENSORFLOW).validate_training_job(profiler_config_parser)
```

Reply:
done.